### DISCUSSION TITLE
Multiple active oauth providers per-user

### DISCUSSION THREAD

**Discussion by LittleBigBug** [directus/directus#9088] [Feature Requests]
I think it would be pretty useful to allow users to 'connect' more than one oAuth provider to one account.

i.e. a website using Directus as a backend can allow users to sign up with username and password, but maybe some extra functionality could be added to the user's experience if they connected say a Steam/Google/Facebook/etc account as well. 

Likely a second table with provider info and a M2O relationship to users.
Would this be entirely breaking to add? I'm happy to contribute and try to start up a PR for this :)

**Comment by aidenfoxx** [comment]
It's certainly possible. The biggest issue I see is how to handle users with providers that require the user is not able to log in through other means, such as LDAP or Kerberos. You'd need to have checks to make sure that a user can't just link their facebook to bypass those.

You might also have to consider how it would work with dynamic registration. Right now for OpenID we use the users `sub` as the primary identifier, which will differ between auth providers. That might be best to just ignore initially though, and just have the dynamic registration create a second account. 🤷

**Reply by LittleBigBug** [reply]
I mean, if they registered with google/fb without being logged in to their existing account, if any existing email on the providers match it can tell them to log in first, I'm sure that multiple emails conflicting anyway would be an issue. In any case maybe registering through OpenID/OAuth should tell the user it's creating them a new account and if they want to proceed first? [Tying in with the feature that will allow people to create accounts with oAuth]

With the LDAP issue, maybe there can be whitelists for what oAuth providers you're allowed to log in with at all, and only allowed to connect them to an existing account. And/or a setting that only allows logging in with the provider you created your account with.

**Reply by aidenfoxx** [reply]
It's a lot to think about since auth providers can differ so much. It was something I had trouble with when I reworked the auth system to be more modular.

Another solution could be to create a new auth driver that is a sort of "OAuth aggregator" allowing the configuration of multiple OAuth services under a single auth driver, since then you wouldn't have to care about how this would affect other providers.

**Reply by rijkvanzanten** [reply]
@aidenfoxx Maybe we could add a flag to the oAuth driver that skips the `external_id`, `auth_data` part, and `provider` in favor of only matching on the email address? That way, you could mix and match as well 🤔

**Reply by aidenfoxx** [reply]
@rijkvanzanten Since we check in `AuthenticationService` that the user returned by the driver matches the authenticating `provider`, we'd have to add driver specific logic into the core authentication code. Not sure how I feel about that.

Edit: Also, stop calling it oAuth gosh dang it!! 😂

**Reply by aidenfoxx** [reply]
I guess we could make the `provider` field nullable, and ignore it in that instance. Still not sure on doing it this way though.

**Reply by rijkvanzanten** [reply]
> Since we check in AuthenticationService that the user returned by the driver matches the authenticating provider, we'd have to add driver specific logic into the core authentication code. Not sure how I feel about that.

Hmm yeah good point 🤔 

> Edit: Also, stop calling it oAuth gosh dang it!! 😂

lMfao eVery tIme 

**Comment by br41nslug** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
